File: android/app/src/main/java/ae/findmybay/feature/auth/AuthViewModel.kt
package ae.findmybay.feature.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import ae.findmybay.data.repo.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AuthUiState(
    val phone: String = "",
    val code: String = "",
    val challengeId: String? = null,
    val resendInSec: Int = 0,
    val loading: Boolean = false,
    val error: String? = null,
    val otpSent: Boolean = false,
    val verified: Boolean = false,
)

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val auth: AuthRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(AuthUiState())
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    fun onPhoneChange(s: String) =
        _state.update { it.copy(phone = sanitizePhone(s), error = null) }

    fun onCodeChange(s: String) =
        _state.update { it.copy(code = s.filter(Char::isDigit).take(6), error = null) }

    /** Validates a UAE-style phone number, then requests an OTP. */
    fun requestOtp(onSent: (String) -> Unit) {
        val phone = _state.value.phone
        if (!isLikelyUaePhone(phone)) {
            _state.update { it.copy(error = "Enter a valid UAE phone in international format, e.g. +971501234567") }
            return
        }
        _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { auth.requestOtp(phone) }
                .onSuccess { challengeId ->
                    _state.update {
                        it.copy(
                            loading = false,
                            otpSent = true,
                            challengeId = challengeId,
                            resendInSec = 30,
                        )
                    }
                    onSent(phone)
                }
                .onFailure { e ->
                    _state.update { it.copy(loading = false, error = friendly(e)) }
                }
        }
    }

    fun verifyOtp(onVerified: () -> Unit) {
        val s = _state.value
        val challengeId = s.challengeId ?: return
        if (s.code.length != 6) {
            _state.update { it.copy(error = "Enter the 6-digit code") }
            return
        }
        _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { auth.verifyOtp(s.phone, challengeId, s.code) }
                .onSuccess {
                    _state.update { it.copy(loading = false, verified = true) }
                    onVerified()
                }
                .onFailure { e ->
                    _state.update { it.copy(loading = false, error = friendly(e)) }
                }
        }
    }

    private fun friendly(e: Throwable): String =
        e.message?.takeIf { it.isNotBlank() } ?: "Something went wrong, please try again."

    /** Strips spaces/dashes; ensures leading '+'. */
    private fun sanitizePhone(s: String): String {
        val cleaned = s.trim().replace(Regex("[\\s\\-()]+"), "")
        return when {
            cleaned.startsWith("+") -> cleaned
            cleaned.startsWith("00") -> "+" + cleaned.drop(2)
            cleaned.startsWith("0") -> "+971" + cleaned.drop(1)   // local UAE format
            cleaned.startsWith("971") -> "+$cleaned"
            cleaned.isEmpty() -> ""
            else -> "+$cleaned"
        }
    }

    private fun isLikelyUaePhone(s: String): Boolean {
        // Permissive: +<country><digits> with at least 9 digits after +. Real validation lives server-side.
        return Regex("^\\+\\d{9,15}$").matches(s)
    }
}

File: ios/FindMyBay/Features/Auth/AuthViewModel.swift
// AuthViewModel.swift
//
// Phone-OTP screen state machine. Mirrors `feature/auth/AuthViewModel.kt`
// — phone sanitization, UAE-style validation, request/verify side-effects.
//
// Uses Swift's @Observable so the view re-renders without needing to wrap
// state in @Published. Equivalent of Compose's `collectAsStateWithLifecycle`.

import Foundation
import Observation

@Observable @MainActor
final class AuthViewModel {

    // MARK: - UI state (matches Kotlin's AuthUiState)
    var phone: String = ""
    var code: String = ""
    var challengeId: String? = nil
    var resendInSec: Int = 0
    var loading: Bool = false
    var error: String? = nil
    var otpSent: Bool = false
    var verified: Bool = false

    // MARK: - Deps
    private let auth: AuthRepository

    init(auth: AuthRepository) {
        self.auth = auth
    }

    // MARK: - Input

    func onPhoneChange(_ s: String) {
        phone = Self.sanitizePhone(s)
        error = nil
    }

    func onCodeChange(_ s: String) {
        code = String(s.filter(\.isNumber).prefix(6))
        error = nil
    }

    // MARK: - Actions

    /// Validates the phone, requests an OTP. Calls `onSent(phone)` on success.
    func requestOtp(onSent: @escaping (String) -> Void) {
        guard Self.isLikelyUaePhone(phone) else {
            error = "Enter a valid UAE phone in international format, e.g. +971501234567"
            return
        }
        loading = true
        error = nil
        Task {
            do {
                let challengeId = try await auth.requestOtp(phone: phone)
                self.loading = false
                self.otpSent = true
                self.challengeId = challengeId
                self.resendInSec = 30
                onSent(self.phone)
            } catch {
                self.loading = false
                self.error = Self.friendly(error)
            }
        }
    }

    /// Drives the Google Sign-In flow then POSTs the ID token to the backend.
    /// Calls `onVerified()` on success — same exit point as OTP verify so
    /// the host's navigation logic doesn't need to differentiate.
    func signInWithGoogle(onVerified: @escaping () -> Void) {
        loading = true
        error = nil
        Task {
            NSLog("FmbAuthDebug GoogleSignIn step=start")
            do {
                let google = try await GoogleSignInService.signIn()
                NSLog("FmbAuthDebug GoogleSignIn step=sdk_returned email=%@ idTokenLen=%d",
                      google.email ?? "nil", google.idToken.count)
                _ = try await auth.signInWithGoogle(idToken: google.idToken)
                NSLog("FmbAuthDebug GoogleSignIn step=backend_ok")
                self.loading = false
                self.verified = true
                onVerified()
            } catch GoogleSignInError.userCancelled {
                NSLog("FmbAuthDebug GoogleSignIn step=cancelled")
                self.loading = false
            } catch let GoogleSignInError.notConfigured(why) {
                NSLog("FmbAuthDebug GoogleSignIn step=not_configured reason=%@", why)
                self.loading = false
                self.error = "Google Sign-In not configured: \(why)"
            } catch let GoogleSignInError.noPresenter {
                NSLog("FmbAuthDebug GoogleSignIn step=no_presenter")
                self.loading = false
                self.error = "Could not present Google Sign-In."
            } catch let GoogleSignInError.noIDToken {
                NSLog("FmbAuthDebug GoogleSignIn step=no_idtoken")
                self.loading = false
                self.error = "Google did not return an ID token."
            } catch let GoogleSignInError.other(inner) {
                NSLog("FmbAuthDebug GoogleSignIn step=sdk_failed err=%@", inner.localizedDescription)
                self.loading = false
                self.error = inner.localizedDescription
            } catch {
                NSLog("FmbAuthDebug GoogleSignIn step=other_failed err=%@", error.localizedDescription)
                self.loading = false
                self.error = Self.friendly(error)
            }
        }
    }

    /// Drives the Sign in with Apple flow then POSTs the identity token to
    /// the backend. Same exit-point contract as the Google + OTP flows so
    /// the host's navigation logic stays uniform.
    func signInWithApple(onVerified: @escaping () -> Void) {
        loading = true
        error = nil
        Task {
            NSLog("FmbAuthDebug AppleSignIn step=start")
            do {
                let apple = try await AppleSignInService.signIn()
                NSLog("FmbAuthDebug AppleSignIn step=sdk_returned email=%@ idTokenLen=%d hasName=%@",
                      apple.email ?? "nil", apple.idToken.count,
                      apple.fullName == nil ? "no" : "yes")
                _ = try await auth.signInWithApple(
                    idToken: apple.idToken,
                    nonce: apple.nonce,
                    fullName: apple.fullName
                )
                NSLog("FmbAuthDebug AppleSignIn step=backend_ok")
                self.loading = false
                self.verified = true
                onVerified()
            } catch AppleSignInError.userCancelled {
                NSLog("FmbAuthDebug AppleSignIn step=cancelled")
                self.loading = false
            } catch let AppleSignInError.other(inner) {
                NSLog("FmbAuthDebug AppleSignIn step=sdk_failed err=%@", inner.localizedDescription)
                self.loading = false
                self.error = inner.localizedDescription
            } catch {
                NSLog("FmbAuthDebug AppleSignIn step=other_failed err=%@", error.localizedDescription)
                self.loading = false
                self.error = Self.friendly(error)
            }
        }
    }

    /// Verifies the entered code. Calls `onVerified()` on success.
    func verifyOtp(onVerified: @escaping () -> Void) {
        guard let challengeId else { return }
        guard code.count == 6 else {
            error = "Enter the 6-digit code"
            return
        }
        loading = true
        error = nil
        Task {
            do {
                _ = try await auth.verifyOtp(phone: phone, challengeId: challengeId, code: code)
                self.loading = false
                self.verified = true
                onVerified()
            } catch {
                self.loading = false
                self.error = Self.friendly(error)
            }
        }
    }

    // MARK: - Static helpers (parity with the Kotlin private fns)

    /// Strips spaces / dashes / parens; ensures a leading `+`. Mirrors
    /// `sanitizePhone` in the Kotlin VM exactly.
    static func sanitizePhone(_ s: String) -> String {
        let cleaned = s
            .trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: "[\\s\\-()]+", with: "", options: .regularExpression)

        if cleaned.isEmpty                     { return "" }
        if cleaned.hasPrefix("+")              { return cleaned }
        if cleaned.hasPrefix("00")             { return "+" + cleaned.dropFirst(2) }
        if cleaned.hasPrefix("0")              { return "+971" + cleaned.dropFirst() }
        if cleaned.hasPrefix("971")            { return "+" + cleaned }
        return "+" + cleaned
    }

    /// Permissive sanity check — `+<country><digits>` with 9–15 digits after the
    /// `+`. Real validation happens server-side. Mirrors `isLikelyUaePhone`.
    static func isLikelyUaePhone(_ s: String) -> Bool {
        s.range(of: #"^\+\d{9,15}$"#, options: .regularExpression) != nil
    }

    static func friendly(_ e: Error) -> String {
        let msg = (e as? LocalizedError)?.errorDescription ?? e.localizedDescription
        return msg.isEmpty ? "Something went wrong, please try again." : msg
    }
}
